File: 0198-house-robber/0198-house-robber.java
class Solution {
    public int rob(int[] nums) {
        int n = nums.length;
        int[] dp = new int[n];
        for(int i=0; i<n; i++) {
            dp[i] = -1;
        }
        return robMoney(n-1, nums, dp);
    }
    public int robMoney(int ind, int[] nums, int[] dp) {
        if(ind == 0) return nums[0];
        if(ind < 0) return 0;

        if(dp[ind] != -1) return dp[ind];

        int pick = nums[ind] + robMoney(ind - 2, nums, dp);
        int nonPick = robMoney(ind - 1, nums, dp);

        return dp[ind] = Math.max(pick, nonPick);

    }
}

File: 0198-house-robber/0198-house-robber.py
class Solution:
    def rob(self, nums: List[int]) -> int:
        n = len(nums)
        dp = [nums[0]]

        for i in range(1, n):
            pick = nums[i]
            if i > 1:
                pick += dp[i-2]
            nonPick = dp[i-1]

            dp.append(max(pick, nonPick))

        return dp[n-1]